### DISCUSSION TITLE
Start Activity while exiting a state and stop it until re-entering the same state

### DISCUSSION THREAD

**Discussion by dsathyakumar** [statelyai/xstate#2087] [Q&A] [answered]
Hello 👋 
I'm not sure if this is something not per spec. If so, please let me know.

Currently activities are started when entering a state and stopped when exiting the state.
I'd like to be able to trigger activities that occur while exiting a state and stopping it when re-entering the same state once more.

Is this possible with xstate ?

**Comment by davidkpiano** [comment, marked as answer]
Activities are deprecated in favor of `invoke` instead.

Can you share some code so that we can get a better idea of what you're trying to do?

**Reply by dsathyakumar** [reply]
Thanks for the response @davidkpiano 👍 
Well, we have an uploader widget that performs _csv_ file uploads to an end point & returns a **_taskId_**.
Then the file is processed async in the background.
From the front end, we use the initially retrieved **taskId** and poll repeatedly showing progress of processing etc., + a ton of status messages.
So, the uploader has states like **idle**, **upload**, **poll**. 

The transition is usually
- idle -> upload -> poll -> idle

We want a universal logger implemented within the machine to keep track of states, events, and some props on context and finally dispatch the logs to our tracking system.
It would be great if this logging began when exiting the `idle` state and stopping when it re-entered `idle`

**Reply by davidkpiano** [reply]
It seems like `poll` can be a child state of `upload` since polling happens during uploading if I understand correctly, right?

> We want a universal logger implemented within the machine to keep track of states, events, and some props on context and finally dispatch the logs to our tracking system.

You can subscribe to the service to have these "general" actions:

```js
service.subscribe(state => {
  if (!state.matches('idle')) {
    console.log(state); // or however you want to log it
  }
});
```
